DESIGN.md

```diff
@@ -0,0 +1,133 @@
+
+### Catalog
+#### Updating the schema
+We implemented a DDL executor for the ALTER TABLE command that updates the catalog when a table is altered. Scehma information for each table (pointer to the schema) is stored in a column in the catalog tables (`classes_` to be more specific). So after a new schema is generated by evaluating the ALTER TABLE commands, we update the schema pointer column for the table being altered. 
+We also add a `layout_version` column along side the schema pointer column for each table. 
+Since the catalog tables are transactional, they disallow write-to-write conflict from MVCC. As a result, concurrent schema updates are caught here as one of the updating transactional will abort when trying to modify the catalog table. 
+
+#### Getting the layout version
+For each transaction that accesses the SqlTable, it needs to know which layout version it should be looking at. We leverage on the transactional tables in the catalog to translate the correct visible layout version for an in-flight transaction. Transaction will read the catalog tables to retrieve the most recent visible layout version, and use it to access the SqlTable.
+
+### ALTERTABLE command (parser, binder, optimizer, planner, and execution)
+ALTER TABLE commands are parsed, binded and optimized just like other commands to generate a plan node tree. As for now, we are only supporting schema updates that do not require scanning of the underlying SqlTable (such as adding columns, dropping columns etc). Such commands only require updating the catalog tables, and the SqlTable's meta data. So a DDL Executor is implemented to execute the generated plan.
+
+
+## Design Rationale
+
+#### Map or vector of datatables? 
+
+Within Sqltable, we can either use a ordered concurrent map of version number to datatable, or use a vector of datatables (index i of vector corresponds to datatable with version number i) augmented with locks.  
+
+We decided to use a fixed-size vector, where the datatable with version i is stored in the i-th position of the vector. We use a MAX_NUM_OF_VERSIONS as the size of the vector, which is the max allowed number of versions for a single datatable.
+
+We use a fixed-size vector because of low synchronization overhead: there is no need for locking the vector since different versions access different indices. Also, we need not make a special case for tables with a single version, because accessing an index in the vector is as fast as reading a pointer. Using a vector is thus much faster than a concurrent map.
+
+One potential issue with fixed-size vector is dealing with datatables with more than MAX_NUM_OF_VERSIONS. This should be extremely rare if we set MAX_NUM_OF_VERSIONS, since schema changes generally don't happen too often on a datatable. When we support GC of stale versions, we can implement recycling segments of the vector with stale versions, and adding offset to allow using this vector in a cyclic fashion,
```

Comment:
Something to consider is that you can reserve MAX_NUM_OF_VERSIONS, which means that the vector is at least of this size. Then, if you do need to dynamically resize, it would be a constant operation amortized, which could be reasonable until there is GC of stale versions.

Reply:
This would be hard to pull off, because vector.resize() is not thread safe, it might move the vector to another memory location, so concurrent accesses on vector locations will be unsafe

Reply:
Resize doesn't need to be called. Reserve just ensures that the size of the vector is at least of a certain size. As for data races with reallocation, "if reallocation happens, the container and all its contained elements are modified. Otherwise, the container is accessed, but not the contained elements: concurrently accessing or modifying them is safe." Source: http://www.cplusplus.com/reference/vector/vector/reserve/

Reply:
well reserving space for the vector is not that much different from our current approach, which basically allocates vector of size MAX_NUM_VERSIONS upfront. I don't see why this is necessary.